pandas/core/generic.py

```diff
@@ -3439,6 +3439,8 @@ def where(self, cond, other=np.nan, inplace=False, axis=None, level=None,
               try_cast=False, raise_on_error=True):
 
         if isinstance(cond, NDFrame):
+            if isinstance(cond, pd.Series) and isinstance(self, pd.DataFrame):
+                cond = self._getitem_array(cond)
```

Comment:
I think would be better to make these align properly, e.g.

``cond, _ = cond.align(self)``

(which currently will raise)
and it will do the ``.reindex``

then inside of ``align`` you can so somethign like that